Fix percent/calculated logical widths for RenderReplaced items with fixed logical heights and intrinsic aspect ratio.

yulunwu2 opened the pull request:
#### bd8721b5d09d4fb5f600b4adb81c7ac6e1240591
<pre>
 Fix percent/calculated logical widths for RenderReplaced items with fixed logical heights and intrinsic aspect ratio.
<a href="https://bugs.webkit.org/show_bug.cgi?id=292568">https://bugs.webkit.org/show_bug.cgi?id=292568</a>
&lt;<a href="https://rdar.apple.com/146172341">rdar://146172341</a>&gt;

Reviewed by Alan Baradlay.

RenderReplaced::computePreferredLogicalWidths() now uses aspect ratio adjusted widths
when fixed height/min-height/max-height is specified. This ensures that intrinsic width
calculations keep the aspect ratio of RenderReplaced elements when vertical constraints
are present.

Combined changes:
* LayoutTests/fast/css/max-content-max-height-img-sizing-expected.html: Added.
* LayoutTests/fast/css/max-content-max-height-img-sizing.html: Added.
* Source/WebCore/rendering/RenderReplaced.cpp:
* Source/WebCore/rendering/RenderReplaced.h:

Canonical link: <a href="https://commits.webkit.org/294735@main">https://commits.webkit.org/294735@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b5b2e4e50ff4a4e673c716bac86a7dcd60e4caa2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102898 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22572 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12892 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108064 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/53539 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/104937 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22892 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31073 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/78232 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/53539 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105904 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17728 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/92857 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/58566 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/17567 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10893 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52896 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/87376 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10960 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110439 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30035 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/22124 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/87218 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30399 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89053 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86837 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22104 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31683 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/9400 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/24297 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29962 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/35284 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29770 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33097 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/31332 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderReplaced.cpp:
```diff
@@ -737,7 +737,25 @@ LayoutUnit RenderReplaced::computeReplacedLogicalHeight(std::optional<LayoutUnit
+        || style().logicalMaxHeight().isSpecified()));
+    };
+    if (!canResolvePercentWidthFromAspectRatio()) {
+        minLogicalWidth = maxLogicalWidth = intrinsicLogicalWidth() / 2;
```
`/ 2` ?

yulunwu2 replied:
Removed. Not sure how that made it there.

alanbaradlay commented on LayoutTests/fast/css/max-content-img-sizing-expected.html:
```diff
@@ -0,0 +1,10 @@
+      width: max-content;
+   }
+</style>
+<div><img src="resources/greenbox.png"></div>
```
the expected case should be considerably different to make sure we don't run the same codepath for, in this case box sizing (it helps to prevent accidental matches, e.g. when a bug in max-content codepath would not show as both tests run into the same problem). In many cases we just use fixed sizing for the expected case to keep things simple.

alanbaradlay commented on Source/WebCore/rendering/RenderReplaced.cpp:
```diff
@@ -738,6 +738,24 @@ LayoutUnit RenderReplaced::computeReplacedLogicalHeight(std::optional<LayoutUnit
+        return;
+
+    if (style().logicalHeight().isFixed())
+        minLogicalWidth = maxLogicalWidth = style().logicalHeight().value();
```
is this correct for all cases? it certainly works for aspect ratio of 1.

yulunwu2 replied:
Christ almighty, fixed

alanbaradlay commented on Source/WebCore/rendering/RenderReplaced.cpp:
```diff
@@ -738,6 +738,24 @@ LayoutUnit RenderReplaced::computeReplacedLogicalHeight(std::optional<LayoutUnit
+
+    auto canResolvePercentWidthFromAspectRatio = [&]() {
+        return (RenderObject::hasIntrinsicAspectRatio()
+        &&(style().logicalHeight().isFixed()
```
`&& (`

yulunwu2 replied:
Removing this block

alanbaradlay commented on Source/WebCore/rendering/RenderReplaced.cpp:
```diff
@@ -738,6 +738,24 @@ LayoutUnit RenderReplaced::computeReplacedLogicalHeight(std::optional<LayoutUnit
+    if (style().logicalHeight().isFixed())
+        minLogicalWidth = maxLogicalWidth = style().logicalHeight().value();
+    else {
+        if (style().logicalMinHeight().isFixed())
```
min/max values still apply when height is fixed (should not be in an else branch)

yulunwu2 replied:
Fixed.

alanbaradlay commented on Source/WebCore/rendering/RenderReplaced.cpp:
```diff
@@ -738,6 +738,24 @@ LayoutUnit RenderReplaced::computeReplacedLogicalHeight(std::optional<LayoutUnit
+        if (style().logicalMinHeight().isFixed())
+            minLogicalWidth = style().logicalMinHeight().value() * computeIntrinsicAspectRatio();
+        if (style().logicalMaxHeight().isFixed())
+            maxLogicalWidth = style().logicalMaxHeight().value() * computeIntrinsicAspectRatio();
```
Wouldn't we want the min value adjusted too? we could easily end up with min > max here (applies to the `min` above too)

yulunwu2 replied:
Fixed.

alanbaradlay commented on Source/WebCore/rendering/RenderReplaced.cpp:
```diff
@@ -738,6 +738,24 @@ LayoutUnit RenderReplaced::computeReplacedLogicalHeight(std::optional<LayoutUnit
+        if (style().logicalMaxHeight().isFixed())
+            maxLogicalWidth = style().logicalMaxHeight().value() * computeIntrinsicAspectRatio();
+        }
 }
```
I think this should be along the lines of

```
void RenderReplaced::computeIntrinsicLogicalWidths(LayoutUnit& minLogicalWidth, LayoutUnit& maxLogicalWidth) const
{
    minLogicalWidth = intrinsicLogicalWidth();
    maxLogicalWidth = minLogicalWidth;

    if (!hasIntrinsicAspectRatio())
        return;

    auto& style = this->style();
    auto computedAspectRatio = computeIntrinsicAspectRatio();
    auto computedIntrinsicLogicalWidth = minLogicalWidth;

    if (style.logicalHeight().isFixed())
        computedIntrinsicLogicalWidth = style.logicalHeight().value() * computedAspectRatio;

    if (style.logicalMaxHeight().isFixed())
        computedIntrinsicLogicalWidth = std::min(computedIntrinsicLogicalWidth, LayoutUnit { style.logicalMaxHeight().value() * computedAspectRatio });

    if (style.logicalMinHeight().isFixed())
        computedIntrinsicLogicalWidth = std::max(computedIntrinsicLogicalWidth, LayoutUnit { style.logicalMinHeight().value() * computedAspectRatio });

    minLogicalWidth = computedIntrinsicLogicalWidth;
    maxLogicalWidth = minLogicalWidth;
}

```
or something similar in functionality.

I'd also add tons of test cases with different min/max/height values to test correctness. The current PR would fail on most variants.

alanbaradlay replied:
-also this logic may belong to `RenderReplaced::computePreferredLogicalWidths` as I think, strictly speaking `computeIntrinsicLogicalWidths` should just return whatever the intrinsic value is and it's the preferred width that is supposed to constrain that value.

yulunwu2 replied:
Updated computePreferredLogicalWidths to call a new function, computeAspectRatioAdjustedIntrinsicLogicalWidths, and reverted computeIntrinsicLogicalWidths.

Adding a handful of test cases.

alanbaradlay commented on LayoutTests/fast/css/max-content-img-sizing.html:
```diff
@@ -0,0 +1,66 @@
+      width: 50%;
+      min-height: 50px;
+}
+</style>
```
could you add cases where both min and max are applied?

yulunwu2 replied:
Done

alanbaradlay replied:
also please make sure content fits the viewport. AFAIR WKTR uses 800x600, but you can check by making the test fail (and checking the screenshot diffs)

yulunwu2 replied:
Yup, the bottom portion was outside the viewport. Splitting this into two tests.

alanbaradlay commented on Source/WebCore/rendering/RenderReplaced.h:
```diff
@@ -59,6 +59,7 @@ class RenderReplaced : public RenderBox {
     void layout() override;
 
     void computeIntrinsicLogicalWidths(LayoutUnit& minLogicalWidth, LayoutUnit& maxLogicalWidth) const final;
+    void computeAspectRatioAdjustedIntrinsicLogicalWidths(LayoutUnit& minLogicalWidth, LayoutUnit& maxLogicalWidth) const;
```
I don't think this needs to be a protected function.

yulunwu2 replied:
Moved here and in .cpp

alanbaradlay replied:
I think having it as private is the most reasonable choice here. I should have been more clear, sorry -also while you are in there, please mark it const. Thank you.

yulunwu2 replied:
Done.

alanbaradlay commented on Source/WebCore/rendering/RenderReplaced.cpp:
```diff
@@ -737,14 +737,37 @@ void RenderReplaced::computeIntrinsicLogicalWidths(LayoutUnit& minLogicalWidth,
+    if (style.logicalMinHeight().isFixed())
+        computedIntrinsicLogicalWidth = std::max(computedIntrinsicLogicalWidth, LayoutUnit { style.logicalMinHeight().value() * computedAspectRatio });
+
+    minLogicalWidth = maxLogicalWidth = computedIntrinsicLogicalWidth;
```
this is against webkit coding style afair.

yulunwu2 replied:
Fixed

yulunwu2 committed:
Fix percent/calculated logical widths for RenderReplaced items with fixed logical heights and intrinsic aspect ratio.
https://bugs.webkit.org/show_bug.cgi?id=292568
<rdar://146172341>

Reviewed by Alan Baradlay.

RenderReplaced::computePreferredLogicalWidths() now uses aspect ratio adjusted widths
when fixed height/min-height/max-height is specified. This ensures that intrinsic width
calculations keep the aspect ratio of RenderReplaced elements when vertical constraints
are present.

Combined changes:
* LayoutTests/fast/css/max-content-max-height-img-sizing-expected.html: Added.
* LayoutTests/fast/css/max-content-max-height-img-sizing.html: Added.
* Source/WebCore/rendering/RenderReplaced.cpp:
* Source/WebCore/rendering/RenderReplaced.h:

Canonical link: https://commits.webkit.org/294735@main
